include/bluetooth/bluetooth.h

```diff
@@ -45,6 +45,75 @@ extern "C" {
  */
 #define BT_ID_DEFAULT 0
 
+/** Opaque type representing an advertiser. */
+struct bt_le_ext_adv;
+
+/* Don't require everyone to include conn.h */
+struct bt_conn;
+
+struct bt_le_adv_sent_info {
+	/* The number of advertising events completed. */
+	u8_t num_sent;
+};
+
+struct bt_le_adv_connected_info {
+	/* Connection object of the new connection */
+	struct bt_conn *conn;
+};
+
+struct bt_le_adv_scanned_info {
+	/* The Bluetooth device address of the active scanner */
+	bt_addr_le_t *addr;
+};
+
+struct bt_le_adv_cb {
+	/** @brief The advertising set has finished sending adv data.
+	 *
+	 *  This callback notifies the application that the advertising set has
+	 *  finished sending advertising data.
+	 *  The advertising set can either have been stopped by a timeout or
+	 *  because the specified number of advertising events has been reached.
+	 *
+	 *  @param adv  The advertising set object.
+	 *  @param info Information about the sent event.
+	 */
+	void (*sent)(struct bt_le_ext_adv *adv,
+		     struct bt_le_adv_sent_info *info);
+
+	/** @brief The advertising set has accepted a new connection.
+	 *
+	 *  This callback notifies the application that the advertising set has
+	 *  accepted a new connection.
+	 *
+	 *  @param adv  The advertising set object.
+	 *  @param info Information about the connected event.
+	 */
+	void (*connected)(struct bt_le_ext_adv *adv,
+			  struct bt_le_adv_connected_info *info);
+
+	/** @brief The advertising set has been scanned.
+	 *
+	 *  This callback notifies the application that the advertising set has
+	 *  been scanned by an active scanner.
+	 *
+	 *  @param adv  The advertising set object.
+	 *  @param addr Information about the scanned event.
+	 */
+	void (*scanned)(struct bt_le_ext_adv *adv,
+			struct bt_le_adv_scanned_info *info);
+
+	sys_snode_t node;
+};
+
+/** @brief Register advertiser callbacks.
+ *
+ *  Adds the callback structure to the list of callback structures that monitors
+ *  advertiser activity.
+ *
+ *  @param cb Callback struct. Must point to static memory.
+ */
+void bt_le_adv_cb_register(struct bt_le_adv_cb *cb);
```

Comment:
Shouldn't the callbacks be per-`bt_le_ext_adv` instance? This seems especially appropriate when we decided to omit reference counting, i.e. it's assumed that a single module has full ownership of the instance. If that's the case then why would it be appropriate for someone to listen in on actions happening to instances that they don't own?

Reply:
What I'd propose here is that you pass the callback struct to `bt_le_ext_adv_create()`

Reply:
Agreed, but maybe instead to `bt_le_ext_adv_start()` instead. That way the callback can be changed every time we start advertising.

Reply:
My vote would be to register the callbacks in create() as opposed to start(). I don't see a use-case for registering it at the start.